kotlinx-coroutines-core/common/src/EventLoop.common.kt

```diff
@@ -487,48 +489,48 @@ internal abstract class EventLoopImplBase: EventLoopImplPlatform(), Delay {
             _heap = DISPOSED_TASK // never add again to any heap
         }
 
-        override fun toString(): String = "Delayed[nanos=$nanoTime]"
+        override fun toString(): String = "Delayed[timeMark=$timeMark]"
     }
 
     private inner class DelayedResumeTask(
-        nanoTime: Long,
+        timeMark: ComparableTimeMark,
         private val cont: CancellableContinuation<Unit>
-    ) : DelayedTask(nanoTime) {
+    ) : DelayedTask(timeMark) {
         override fun run() { with(cont) { resumeUndispatched(Unit) } }
         override fun toString(): String = super.toString() + cont.toString()
     }
 
     private class DelayedRunnableTask(
-        nanoTime: Long,
+        timeMark: ComparableTimeMark,
         private val block: Runnable
-    ) : DelayedTask(nanoTime) {
+    ) : DelayedTask(timeMark) {
         override fun run() { block.run() }
         override fun toString(): String = super.toString() + block.toString()
     }
 
     /**
      * Delayed task queue maintains stable time-comparision invariant despite potential wraparounds in
-     * long nano time measurements by maintaining last observed [timeNow]. It protects the integrity of the
+     * `nanoTime()` measurements by maintaining last observed [timeMark]. It protects the integrity of the
      * heap data structure in spite of potential non-monotonicity of `nanoTime()` source.
      * The invariant is that for every scheduled [DelayedTask]:
      *
      * ```
-     * delayedTask.nanoTime - timeNow >= 0
+     * delayedTask.timeMark - this.timeMark >= Duration.ZERO
      * ```
      *
      * So the comparison of scheduled tasks via [DelayedTask.compareTo] is always stable as
-     * scheduled [DelayedTask.nanoTime] can be at most [Long.MAX_VALUE] apart. This invariant is maintained when
-     * new tasks are added by [DelayedTask.scheduleTask] function and it cannot be violated when tasks are removed
-     * (so there is nothing special to do there).
+     * scheduled [DelayedTask.timeMark] can be at most [Long.MAX_VALUE] nanoseconds apart.
+     * This invariant is maintained when new tasks are added by [DelayedTask.scheduleTask] function
+     * and it cannot be violated when tasks are removed (so there is nothing special to do there).
      */
     internal class DelayedTaskQueue(
-        @JvmField var timeNow: Long
+        @JvmField var timeMark: ComparableTimeMark
     ) : ThreadSafeHeap<DelayedTask>()
 }
 
 internal expect fun createEventLoop(): EventLoop
 
-internal expect fun nanoTime(): Long
+internal expect fun markNow(): ComparableTimeMark
```

Comment:
just curious: is it possible to use [ValueTimeMark](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.time/-time-source/-monotonic/-value-time-mark/) which is returned by [TimeSource.Monotonic#markNow](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.time/-time-source/-monotonic/) to avoid boxing? And does it make sense?

Reply:
The issue is that only `TimeSource.Monotonic` returns `ValueTimeMark`. For testing purposes, the project uses `TestTimeSource` to generate predictable time marks, which are not `ValueTimeMark`.